### PR TITLE
[GEOT-6895] - Restrict the responsibilities of WMTSTileService

### PR DESCRIPTION
**Pull request by roarbra**
[GEOT-6895](https://osgeo-org.atlassian.net/browse/GEOT-6895) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

The goal of this PR is to clean up the separation between WebMapTileServer and WMTSTileServer. WebMapTileServer will deal with the capabilities document, and give out information about available layers, styles and formats. WMTSTileServer is a subclass of TileServer and should only deal with one single cache of images for a specific layer, style.

I've added some helper classes based on code fragments I found, and to make url building faster. That might be replaced by some existing utility-classes.

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [X] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [X] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [X] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [X] PR for bug fixes and small new features are presented as a single commit
- [X] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [X] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

**Comment by aaime** [comment]
Two easy to fix PMD checks failed:

```
2021-06-03T08:58:13.9515630Z [INFO] PMD Failure: org.geotools.ows.wmts.client.TileURLBuilder:55 Rule:UseDiamondOperator Priority:3 Explicit type arguments can be replaced by Diamond Operator.
2021-06-03T08:58:13.9525082Z [INFO] PMD Failure: org.geotools.ows.wmts.WMTSHelperTest:68 Rule:UseDiamondOperator Priority:3 Explicit type arguments can be replaced by Diamond Operator.
```

File: modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+/**
+ * Some useful functions that was used within gt-wmts.
+ *
+ * <p>Might have their substitution elsewhere
+ *
+ * @author Roar Brænden
+ */
+public class WMTSHelper {
+
+    public static String replaceToken(String baseUrl, String dimName, String dimValue) {
```
Comment (aaime) [inline modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java:32]: Javadoc on public members. I was going to suggest to make the class package private, but it's used in two other packages, so not an avenue

File: modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    /**
+     * Append the parameters to the baseUrl.
+     *
+     * <p>Add a ? mark if baseUrl doesn't have it.
+     *
+     * <p>Does not add existing parameters
+     *
+     * @param baseUrl
```
Comment (aaime) [inline modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java:51]: No point having @param and @return if not documented. I often just get rid of them (they are not mandatory)

File: modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                    }
+                }
+            }
+            return baseUrl + arguments.toString();
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException("Doesn't support UTF-8", e);
+        }
+    }
+
+    public static String usePercentEncodingForSpace(String name) {
```
Comment (aaime) [inline modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java:84]: Javadoc on public members.

**Review by aaime — COMMENTED** [review]
Did a quick pass, found a couple of minor things. I have however no experience with the gt-wmts module, would be useful if @ianturton or @etj could have a look too.

File: modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java (hunk 3 above)
Comment (roarbra) [reply modules/extension/wmts/src/main/java/org/geotools/ows/wmts/WMTSHelper.java:84]: Fixed the comments and made one of the functions package private. Would like to eliminate the whole class though. Replacing with some other utility class.

**Comment by roarbra** [comment]
Fixed the comments and made one of the functions package private. Would like to eliminate the whole WMTSHelper class though. Replacing with some other utility class.

**Comment by aaime** [comment]
@roarbra it would be best if WMTSHelper is replaced sooner rather than later. If it gets into a release, then it's public API and removing it will require going through a deprecation cycle and the like. 
How do you want to proceed? I'm also happy to merge now.

**Comment by roarbra** [comment]
Hi,

I haven’t got time to look for a replacement of these helper functions now,
and I would like to work more on the gt-wmts / gt-tile-client
interrelationship during the summer.

So a merge would be much aprecciated.

Regards,

Roar Brænden



fre. 25. jun. 2021 kl. 18:43 skrev Andrea Aime ***@***.***>:

**Comment by aaime** [comment]
Merge it is then.

**Comment by roarbra** [comment]
@aaime Thanks for the merges!

**Comment by aaime** [comment]
Unfortunately this PR caused build failures in all builders (PR checks and on build.geoserver.org):

```
[ERROR] Tests run: 7, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 2.552 s <<< FAILURE! - in org.geotools.ows.wmts.client.WMTSTileFactory4326Test
[ERROR] testCreateTileFromXYZoom(org.geotools.ows.wmts.client.WMTSTileFactory4326Test)  Time elapsed: 0.68 s  <<< FAILURE!
java.lang.AssertionError: expected:<http://demo.geo-solutions.it/geoserver/gwc/service/wmts?REQUEST=getcapabilities&tilematrixset=EPSG%3A4326&TileRow=15&service=WMTS&format=image%2Fpng&TileCol=20&version=1.0.0&layer=unesco%3AUnesco_point&TileMatrix=EPSG%3A4326%3A5> but was:<http://demo.geo-solutions.it/geoserver/gwc/service/wmts?REQUEST=getcapabilities&tilematrixset=EPSG%3A4326&TileRow=15&service=WMTS&format=image%2Fpng&TileCol=20&version=1.0.0&layer=unesco%3AUnesco_point&TileMatrix=EPSG:4326:5>
	at org.geotools.ows.wmts.client.WMTSTileFactory4326Test.testCreateTileFromXYZoom(WMTSTileFactory4326Test.java:95)
```

The new URLBuiler class is not URL encoding the parts it's adding, the tile matrix in WMTS is not necessarily a number.
Since I'm leaving for vacation Monday morning, I have to get the build back on track, cannot leave it this way the whole week.

**Comment by roarbra** [comment]
That was strange. I'm sure I've run those tests.
I will look into it.

**Comment by aaime** [comment]
Yeah and when I merged the PR was green so... something new happened in between.

**Pull request merged**